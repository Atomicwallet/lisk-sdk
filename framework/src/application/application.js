/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const assert = require('assert');
const {
	TransferTransaction,
	SecondSignatureTransaction,
	DelegateTransaction,
	VoteTransaction,
	MultisignatureTransaction,
	transactionInterface,
} = require('@liskhq/lisk-transactions');
const { getNetworkIdentifier } = require('@liskhq/lisk-cryptography');
const { validator: liskValidator } = require('@liskhq/lisk-validator');
const _ = require('lodash');
const Controller = require('../controller/controller');
const version = require('../version');
const validator = require('./validator');
const configurator = require('./default_configurator');
const { genesisBlockSchema, constantsSchema } = require('./schema');

const ApplicationState = require('./application_state');

const { createLoggerComponent } = require('../components/logger');
const { createStorageComponent } = require('../components/storage');
const { MigrationEntity } = require('../application/storage/entities');

const { Network } = require('./network');
const { NetworkInfoEntity } = require('../application/storage/entities');
const { networkMigrations } = require('../application/storage/migrations');

const Node = require('./node');

const { InMemoryChannel } = require('../controller/channels');

const HttpAPIModule = require('../modules/http_api');

const registerProcessHooks = app => {
	process.title = `${app.config.app.label}(${app.config.app.version})`;

	process.on('uncaughtException', err => {
		// Handle error safely
		app.logger.error(
			{
				err,
			},
			'System error: uncaughtException',
		);
		app.shutdown(1, err.message);
	});

	process.on('unhandledRejection', err => {
		// Handle error safely
		app.logger.fatal(
			{
				err,
			},
			'System error: unhandledRejection',
		);
		app.shutdown(1, err.message);
	});

	process.once('SIGTERM', () => app.shutdown(1));

	process.once('SIGINT', () => app.shutdown(1));

	process.once('cleanup', (error, code) => app.shutdown(code, error));

	process.once('exit', (error, code) => app.shutdown(code, error));
};

class Application {
	constructor(genesisBlock, config = {}) {
		const errors = liskValidator.validate(genesisBlockSchema, genesisBlock);
		if (errors.length) {
			throw errors;
		}

		// Don't change the object parameters provided
		let appConfig = _.cloneDeep(config);

		if (!_.has(appConfig, 'app.label')) {
			_.set(
				appConfig,
				'app.label',
				`lisk-${genesisBlock.payloadHash.slice(0, 7)}`,
			);
		}

		if (!_.has(appConfig, 'components.logger.logFileName')) {
			_.set(
				appConfig,
				'components.logger.logFileName',
				`${process.cwd()}/logs/${appConfig.app.label}/lisk.log`,
			);
		}

		appConfig = configurator.getConfig(appConfig, {
			failOnInvalidArg: process.env.NODE_ENV !== 'test',
		});

		// These constants are readonly we are loading up their default values
		// In additional validating those values so any wrongly changed value
		// by us can be catch on application startup
		const constants = validator.parseEnvArgAndValidate(constantsSchema, {});

		// app.genesisConfig are actually old constants
		// we are merging these here to refactor the underlying code in other iteration
		this.constants = { ...constants, ...appConfig.app.genesisConfig };
		this.genesisBlock = genesisBlock;
		this.config = appConfig;
		this.channel = null;
		this.initialState = null;
		this.applicationState = null;

		// TODO: This should be removed after https://github.com/LiskHQ/lisk/pull/2980
		global.constants = this.constants;
		this.logger = this._initLogger();
		this.storage = this._initStorage();

		// Private members
		this._modules = {};
		this._transactions = {};
		this._migrations = {};
		this._node = null;
		this._network = null;
		this._controller = null;

		this.registerTransaction(TransferTransaction);
		this.registerTransaction(SecondSignatureTransaction);
		this.registerTransaction(DelegateTransaction);
		this.registerTransaction(VoteTransaction);
		this.registerTransaction(MultisignatureTransaction);

		this.registerModule(Node, {
			registeredTransactions: this.getTransactions(),
		});
		this.registerModule(HttpAPIModule);
		this.overrideModuleOptions(HttpAPIModule.alias, {
			loadAsChildProcess: true,
		});
	}

	registerModule(moduleKlass, options = {}, alias = undefined) {
		assert(moduleKlass, 'ModuleSpec is required');
		assert(
			typeof options === 'object',
			'Module options must be provided or set to empty object.',
		);
		assert(alias || moduleKlass.alias, 'Module alias must be provided.');
		const moduleAlias = alias || moduleKlass.alias;
		assert(
			!Object.keys(this.getModules()).includes(moduleAlias),
			`A module with alias "${moduleAlias}" already registered.`,
		);

		this.config.modules[moduleAlias] = Object.assign(
			this.config.modules[moduleAlias] || {},
			options,
		);
		this._modules[moduleAlias] = moduleKlass;

		// Register migrations defined by the module
		this.registerMigrations(moduleKlass.alias, moduleKlass.migrations);
	}

	overrideModuleOptions(alias, options) {
		const modules = this.getModules();
		assert(
			Object.keys(modules).includes(alias),
			`No module ${alias} is registered`,
		);
		this.config.modules[alias] = {
			...this.config.modules[alias],
			...options,
		};
	}

	registerTransaction(Transaction, { matcher } = {}) {
		assert(Transaction, 'Transaction implementation is required');

		assert(
			Number.isInteger(Transaction.TYPE),
			'Transaction type is required as an integer',
		);

		assert(
			!Object.keys(this.getTransactions()).includes(
				Transaction.TYPE.toString(),
			),
			`A transaction type "${Transaction.TYPE}" is already registered.`,
		);

		validator.validate(transactionInterface, Transaction.prototype);

		if (matcher) {
			Object.defineProperty(Transaction.prototype, 'matcher', {
				get: () => matcher,
			});
		}

		this._transactions[Transaction.TYPE] = Object.freeze(Transaction);
	}

	registerMigrations(namespace, migrations) {
		assert(namespace, 'Namespace is required');
		assert(Array.isArray(migrations), 'Migrations list should be an array');
		assert(
			!Object.keys(this.getMigrations()).includes(namespace),
			`Migrations for "${namespace}" was already registered.`,
		);

		this._migrations[namespace] = Object.freeze(migrations);
	}

	getTransactions() {
		return this._transactions;
	}

	getTransaction(transactionType) {
		return this._transactions[transactionType];
	}

	getModule(alias) {
		return this._modules.get[alias];
	}

	getModules() {
		return this._modules;
	}

	getMigrations() {
		return this._migrations;
	}

	async run() {
		this.logger.info(
			'If you experience any type of error, please open an issue on Lisk GitHub: https://github.com/LiskHQ/lisk-sdk/issues',
		);
		this.logger.info(
			'Contribution guidelines can be found at Lisk-docs: https://github.com/LiskHQ/lisk-docs/blob/build/CONTRIBUTING.adoc',
		);
		this.logger.info(`Booting the application with Lisk Framework(${version})`);

		// Freeze every module and configuration so it would not interrupt the app execution
		this._compileAndValidateConfigurations();

		Object.freeze(this.genesisBlock);
		Object.freeze(this.constants);
		Object.freeze(this.config);

		this.logger.info(`Starting the app - ${this.config.app.label}`);

		registerProcessHooks(this);

		// Initialize all objects
		this.applicationState = this._initApplicationState();
		this.channel = this._initChannel();
		this.applicationState.channel = this.channel;

		this._controller = this._initController();
		this._network = this._initNetwork();

		// Load system components
		await this.storage.bootstrap();
		await this.storage.entities.Migration.defineSchema();

		await this._controller.load(
			this.getModules(),
			this.config.modules,
			this.getMigrations(),
		);

		await this._network.initialiseNetwork();

		this.channel.publish('app:ready');
	}

	async shutdown(errorCode = 0, message = '') {
		if (this._controller) {
			await this._controller.cleanup(errorCode, message);
		}
		this.logger.info({ errorCode, message }, 'Shutting down application');
		process.exit(errorCode);
	}

	// --------------------------------------
	// Private
	// --------------------------------------
	_compileAndValidateConfigurations() {
		const modules = this.getModules();
		this.config.app.networkId = getNetworkIdentifier(
			this.genesisBlock.payloadHash,
			this.genesisBlock.communityIdentifier,
		);

		const appConfigToShareWithModules = {
			version: this.config.app.version,
			minVersion: this.config.app.minVersion,
			protocolVersion: this.config.app.protocolVersion,
			networkId: this.config.app.networkId,
			genesisBlock: this.genesisBlock,
			constants: this.constants,
			lastCommitId: this.config.app.lastCommitId,
			buildVersion: this.config.app.buildVersion,
		};

		// TODO: move this configuration to module especific config file
		const childProcessModules = process.env.LISK_CHILD_PROCESS_MODULES
			? process.env.LISK_CHILD_PROCESS_MODULES.split(',')
			: [];

		Object.keys(modules).forEach(alias => {
			this.overrideModuleOptions(alias, {
				loadAsChildProcess: childProcessModules.includes(alias),
			});
			this.overrideModuleOptions(alias, appConfigToShareWithModules);
		});

		this.initialState = {
			version: this.config.app.version,
			minVersion: this.config.app.minVersion,
			protocolVersion: this.config.app.protocolVersion,
			networkId: this.config.app.networkId,
			wsPort: this.config.app.network.wsPort,
			httpPort: this.config.modules.http_api.httpPort,
		};

		this.logger.trace(this.config, 'Compiled configurations');
	}

	_initLogger() {
		return createLoggerComponent({
			...this.config.components.logger,
			module: 'lisk:app',
		});
	}

	_initStorage() {
		const storageConfig = this.config.components.storage;
		const loggerConfig = this.config.components.logger;
		const dbLogger =
			storageConfig.logFileName &&
			storageConfig.logFileName === loggerConfig.logFileName
				? this.logger
				: createLoggerComponent({
						...loggerConfig,
						logFileName: storageConfig.logFileName,
						module: 'lisk:app:database',
				  });

		const storage = createStorageComponent(
			this.config.components.storage,
			dbLogger,
		);

		storage.registerEntity('Migration', MigrationEntity);

		return storage;
	}

	_initApplicationState() {
		return new ApplicationState({
			initialState: this.initialState,
			logger: this.logger,
		});
	}

	_initChannel() {
		return new InMemoryChannel(
			'app',
			['ready', 'state:updated'],
			{
				getComponentConfig: {
					handler: action => this.config.components[action.params],
				},
				getApplicationState: {
					handler: () => this.applicationState.state,
				},
				updateApplicationState: {
					handler: action => this.applicationState.update(action.params),
				},
				sendToNetwork: {
					handler: action => this._network.send(action.params),
				},
				broadcastToNetwork: {
					handler: action => this._network.broadcast(action.params),
				},
				requestFromNetwork: {
					handler: action => this._network.request(action.params),
				},
				requestFromPeer: {
					handler: action => this._network.requestFromPeer(action.params),
				},
				getConnectedPeers: {
					handler: action => this._network.getConnectedPeers(action.params),
				},
				getDisconnectedPeers: {
					handler: action => this._network.getDisconnectedPeers(action.params),
				},
				applyPenaltyOnPeer: {
					handler: action => this._network.applyPenalty(action.params),
				},
			},
			{ skipInternalEvents: true },
		);
	}

	_initController() {
		return new Controller({
			appLabel: this.config.app.label,
			config: {
				components: this.config.components,
				ipc: this.config.app.ipc,
				tempPath: this.config.app.tempPath,
			},
			logger: this.logger,
			storage: this.storage,
			channel: this.channel,
		});
	}

	_initNetwork() {
		const network = new Network({
			options: this.config.app.network,
			storage: this.storage,
			logger: this.logger,
			channel: this.channel,
		});

		this.storage.registerEntity('NetworkInfo', NetworkInfoEntity);
		this.registerMigrations('network', networkMigrations());

		return network;
	}
}

module.exports = Application;
