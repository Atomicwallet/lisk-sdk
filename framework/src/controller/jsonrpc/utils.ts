/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { validator, LiskValidationError } from '@liskhq/lisk-validator';
import {
	JsonRpcError,
	ID,
	NotificationRequest,
	JsonRpcResult,
	ResponseObjectWithResult,
	ResponseObjectWithError,
} from './types';

export const VERSION = '2.0';

const RequestSchema = {
	id: 'jsonRPCRequestSchema',
	type: 'object',
	required: ['jsonrpc', 'method'],
	properties: {
		jsonrpc: {
			type: 'string',
		},
		method: {
			type: 'string',
		},
		id: {
			type: ['number', 'string', 'null'], // https://www.jsonrpc.org/specification#id1
		},
		params: {
			type: 'object',
		},
	},
};

export const validateJSONRPCRequest = (data: Record<string, unknown>): void => {
	const errors = validator.validate(RequestSchema, data);
	if (errors.length) {
		throw new LiskValidationError(errors);
	}
};

export const notificationRequest = (
	method: string,
	params?: Record<string, unknown>,
): NotificationRequest => ({
	jsonrpc: VERSION,
	method,
	params,
});

export const successResponse = (id: ID, result: JsonRpcResult): ResponseObjectWithResult => ({
	jsonrpc: VERSION,
	id,
	result,
});

export const errorResponse = (id: ID, error: JsonRpcError): ResponseObjectWithError => ({
	jsonrpc: VERSION,
	id,
	error,
});

export const invalidRequest = (): JsonRpcError => ({ message: 'Invalid request', code: -32600 });

export const methodNotFound = (): JsonRpcError => ({ message: 'Method not found', code: -32601 });

export const invalidParams = (): JsonRpcError => ({ message: 'Invalid params', code: -32602 });

export const internalError = (data?: JsonRpcResult): JsonRpcError => {
	if (data) {
		return { message: 'Internal error', code: -32603, data };
	}
	return { message: 'Internal error', code: -32603 };
};

export const parseError = (): JsonRpcError => ({ message: 'Parse error', code: -32700 });
