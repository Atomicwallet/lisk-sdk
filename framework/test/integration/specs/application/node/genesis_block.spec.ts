/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { KVStore } from '@liskhq/lisk-db';
import { TransferTransaction } from '@liskhq/lisk-transactions';
import { nodeUtils } from '../../../../utils';
import { createDB, removeDB } from '../../../../utils/kv_store';
import { Node } from '../../../../../src/application/node';
import { genesis, genesisBlock as getGenesisBlock } from '../../../../fixtures';

describe('genesis block', () => {
	const dbName = 'genesis_block';
	const genesisBlock = getGenesisBlock();
	let node: Node;
	let blockchainDB: KVStore;
	let forgerDB: KVStore;
	let nodeDB: KVStore;

	beforeAll(async () => {
		({ blockchainDB, forgerDB, nodeDB } = createDB(dbName));
		node = await nodeUtils.createAndLoadNode(blockchainDB, forgerDB, nodeDB);
		await node['_forger'].loadDelegates();
	});

	afterAll(async () => {
		await node.cleanup();
		await blockchainDB.close();
		await forgerDB.close();
		removeDB(dbName);
	});

	describe('given the application has not been initialized', () => {
		describe('when chain module is bootstrapped', () => {
			it('should save genesis block to the database', async () => {
				const block = await node['_chain'].dataAccess.getBlockByID(
					genesisBlock.header.id,
				);

				expect(block.header.version).toEqual(0);
				expect(block.header).toEqual(genesisBlock.header);
			});

			it('should save accounts from genesis block assets', async () => {
				// Get genesis accounts
				const genesisAccounts = genesisBlock.header.asset.accounts;

				// Get delegate accounts in genesis block from the database
				const accountsFromDb = await Promise.all(
					genesisAccounts.map(async account =>
						node['_chain'].dataAccess.getAccountByAddress(account.address),
					),
				);

				expect(genesisAccounts).toEqual(accountsFromDb);
			});

			it('should have correct delegate list', async () => {
				const delegateListFromChain = await nodeUtils.getDelegateList(node, 1);
				expect(delegateListFromChain).toMatchSnapshot();
			});
		});
	});

	describe('given the application was initialized earlier', () => {
		const account =
			genesisBlock.header.asset.accounts[
				genesisBlock.header.asset.accounts.length - 1
			];
		let newBalance: bigint;
		let oldBalance: bigint;

		beforeEach(async () => {
			const genesisAccount = await node[
				'_chain'
			].dataAccess.getAccountByAddress(genesis.address);
			const recipient = await node['_chain'].dataAccess.getAccountByAddress(
				account.address,
			);
			oldBalance = account.balance;
			newBalance = oldBalance + BigInt('100000000000');

			const transaction = new TransferTransaction({
				nonce: genesisAccount.nonce,
				senderPublicKey: genesis.publicKey,
				fee: BigInt('200000'),
				asset: {
					recipientAddress: recipient.address,
					amount: BigInt('100000000000'),
					data: '',
				},
			});
			transaction.sign(node['_networkIdentifier'], genesis.passphrase);
			const newBlock = await nodeUtils.createBlock(node, [transaction]);
			await node['_processor'].process(newBlock);
		});

		describe('when chain module is bootstrapped', () => {
			it('should not apply the genesis block again', async () => {
				// Act
				// Re-initialize the node
				node = await nodeUtils.createAndLoadNode(
					blockchainDB,
					forgerDB,
					nodeDB,
				);

				// Arrange & Assert
				const recipient = await node['_chain'].dataAccess.getAccountByAddress(
					account.address,
				);
				expect(recipient.balance).toEqual(newBalance);
			});
		});
	});
});
