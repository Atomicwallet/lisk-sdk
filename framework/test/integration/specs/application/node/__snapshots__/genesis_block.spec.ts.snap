// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genesis block given the application has been initialized previously when chain module is bootstrapped should have correct delegate list 1`] = `
Array [
  "c98554123062ac5795a3ee905b081e863db5a818",
  "936f3a0f4d776b6a7722ed126e8ff17b44d7e7b8",
  "58d907d26508603e838423daa2061c29c7a84950",
  "ad42f8e867d618171bf4982e64269442148f6e11",
  "6ffcd8ad547d8a549a31b25236e322c781a52d85",
  "0bc3bec2fdb565996fd316e368e66e5d8e830808",
  "1c194c2be1cc53f663a93c64899cbaa34016f415",
  "ca309a5f4bbf11ca86592febb6d2ccc78309f69e",
  "4fd52f67f151fbbdda9dd92a714884a399830eca",
  "acfbdbaeb93d587170c7cd9c0b5ffdeb7ff9daec",
  "9139c91f8a0aa1fb385770feaf299b99883aec2d",
  "436b40f58c0c27ed133f6001a019ff25561efad4",
  "e39316cc020089ea7a5614bcf69a8931c10630a7",
  "52f9cdcff0605241c78278690ae36eb0136a30ff",
  "31204ad5b95dd922c2899aa5bf8e7ee5b7546af3",
  "a28d5e34007fd8fe6d7903044eb23a60fdad3c00",
  "9cabee3d27426676b852ce6b804cb2fdff7cd0b5",
  "ca5f6d76eab6e4f5aacee2864c79034d7111b986",
  "0ada6a2f6c8f891769366fc9aa6fd9f1facb36cf",
  "d3c8064d011ef853e3be506b95a045f41f78e72a",
  "70abf056bd92e8f77cfc551748fa54a4e3018d5f",
  "79f30c1cbc1b9c4949c8b85acc24a7578e01558b",
  "657f610728eef97d55e50212871f0993bb7cc700",
  "a0bc50b27e7ac39060ed015a55f2f4508c84f0c2",
  "65f927187bf96aac5d968fcc9351e5492b5f9356",
  "d2c9a93755aed20c4d8f55c1e92b812d2c7d49d2",
  "5fbd442a4647b079cda1229ecf6d8f44f361c8ca",
  "27843a60a1e044c1e6e3cf119fdf64eb2b3e0d94",
  "3deeb0a7426a028b435b4ddd8d35ac85cf567237",
  "328d0f546695c5fa02105deb055cf2801d9b8ba1",
  "abd2ed5ad35b3a0870aadae6dceacc988ba63895",
  "4e874bcfb6f5896fe9e5dab3b26f59b2e2a9c09b",
  "7d2c6781d873ed2ba7a87f46f735f5e15a41a6f1",
  "df0e187bb3895806261c87cf66e1772566ee8e58",
  "0903f4c5cb599a7928aef27e314e98291d1e3888",
  "61f396d2a4a13ab7a39ba791fac4b921b54a208e",
  "2159f75e5440c36431aedbc7dc29a65a327778b8",
  "2cf52c08cc76091d884e800c1c697b13f69907d4",
  "94146c9889748c7b727eb3ac8c20e53c52effd32",
  "6e12e4498ae69fb07ff2d8aab036a911229d6c62",
  "8b1c221a030cf720736d9fb7d0499dd7276fc1b3",
  "b11c5811ea074a30142d824b6e8cfd3df14b2688",
  "ffce8ce225c5d80098f50e877125b655aef6d101",
  "4b6126597881cb6ba1a45c1f6286769e7a094fb4",
  "0d2c377e936b68c70066613b10c0fdad537f90da",
  "6174515fa66c91bff1128913edd4e0f1de37cee0",
  "8eceffd5a41e678b6467c9bc80ce35d2e8543d98",
  "7d60db187337cbd881140d69d84c9246eda8382e",
  "246fba5c519576d93c5fac899c44b29b72f526ae",
  "a9c66694dd65b2fdf40cdf45a0c308cbd38004fc",
  "8074f0d02f748fc55448a4bf200f1dade8517059",
  "82cbc7b39d35af358f9e2513af13b2f77b647a00",
  "5ade564399e670bd1d429583059067f3a6ca2b7f",
  "9b42e4264020f3c3dcaaed806578ccd469205060",
  "d06fe6d3e5f7facb5855eca839422fe3824a5d6e",
  "3c80e7d9964a1c83a6dd5dc64e105e0e634bd58a",
  "03f6d90b7dbd0497dc3a52d1c27e23bb8c75897f",
  "4fd8cc4e27a3489b57ed986efe3d327d3de40d92",
  "31fe789b43277e35ab410f2afcfb574280af2dd8",
  "fa526a1611ccc66dec815cb963174118074b736e",
  "dcb5bf35b6d521195e613c42483f520139e2331d",
  "b76a0f1819c4be0a1482567ca9b9fbed3eda444c",
  "c3ab2ac23512d9bf62b02775e22cf80df814eb1b",
  "4f4422eb61c45edb4d76f10cd871c9f983f2ebaa",
  "e9355152c117c9e1fad8be86e9abea961cef4a36",
  "b485becd88db1ab3d556d405204451ba00adaa7d",
  "308a95d1d3f7bb556f48da4f4344566e59f6f1cb",
  "b7580969dd56151f608931f126f793bbf45d8fa0",
  "8ac800124d5b16afd57b5cf7245edfcd5885ea3b",
  "f730cb929a1c45032387c345e10d2427bea55a5e",
  "290abc4a2244bf0ecf5aa1ccee8ac8f60f8bce48",
  "463e7e879b7bdc6a97ec02a2a603aa1a46a04c80",
  "5cd1d0ccf98f2bd5a4bfaa770d55f16498af0bcc",
  "e2950a9f07b44e724df2129360cc140293c08308",
  "ab0041a7d3f7b2c290b5b834d46bdc7b7eb85815",
  "a6f6a0543ae470c6b056021cb2ac153368eafeec",
  "8506f3c10f75044946f1a23a7caf578253649471",
  "d5bd2050b74b309d54819ca17add173c6fca1e16",
  "8459b8870fcefff59f172d716b7bfe9fcc28d408",
  "89b144ecfdd5ea352083bf624d3cf842ec06a5e3",
  "be89f4e983dfb04e2b58a12eb9ed18149e108b07",
  "d5c4e380b1ec2f7f2068cfba9a90cb3ae7816110",
  "5853a3f24990deecced49d6bc15990102ec0c33a",
  "cb579ee537b34926d47129a0b54c0e6d00ef3004",
  "a0620472cde03e77caece701ab7bc5928a5d367c",
  "d5e1f52cbe4a11a3730b98f52109b57602a9c4a1",
  "aebd99f07218109162a905d0e0c91e58bedc83c5",
  "d8e611bafd70a549f035cf61ab0d6ed9e7f25c4e",
  "0f33a5033b750e6c4dca47e38ba020e912df143e",
  "8722453383f781d5427a4ee211020e49bf34a2b9",
  "bd175729d4177259c71cf13fd4ecfb5d01542706",
  "6330fd8ae91df4a5d7fbc2390c182ec6676dc5a6",
  "3b3e137b1bec6f20c9a8b2ad4f5784661fb0fa79",
  "3de95e18f18a54e2269bbf8f1a38ea70762c73fa",
  "c697b620c7c4015e32dd7bdd7d0430b33404e107",
  "d0a0e45b950e3871d8783b973409042b4ab382d4",
  "9d0149b0962d44bfc08a9f64d5afceb6281d7fb5",
  "24c130eb6cc0d8f663a8f6d16ffc61f935a2e02e",
  "1ac73bff74924ad9b74236c4962be27174ae87d0",
  "3b96d8565569421f43684b2c4eaa0639cbb5e011",
  "b56c55b9a70c8e2f07979b862374aed0e92a6dda",
  "b543e2e592200beb38235f6e48f8abe1d87ad872",
  "6b9895c31bcdb2d9c929b9da7e389ed91de672a0",
]
`;

exports[`genesis block given the application has not been initialized when chain module is bootstrapped should have correct delegate list 1`] = `
Array [
  "c98554123062ac5795a3ee905b081e863db5a818",
  "936f3a0f4d776b6a7722ed126e8ff17b44d7e7b8",
  "58d907d26508603e838423daa2061c29c7a84950",
  "ad42f8e867d618171bf4982e64269442148f6e11",
  "6ffcd8ad547d8a549a31b25236e322c781a52d85",
  "0bc3bec2fdb565996fd316e368e66e5d8e830808",
  "1c194c2be1cc53f663a93c64899cbaa34016f415",
  "ca309a5f4bbf11ca86592febb6d2ccc78309f69e",
  "4fd52f67f151fbbdda9dd92a714884a399830eca",
  "acfbdbaeb93d587170c7cd9c0b5ffdeb7ff9daec",
  "9139c91f8a0aa1fb385770feaf299b99883aec2d",
  "436b40f58c0c27ed133f6001a019ff25561efad4",
  "e39316cc020089ea7a5614bcf69a8931c10630a7",
  "52f9cdcff0605241c78278690ae36eb0136a30ff",
  "31204ad5b95dd922c2899aa5bf8e7ee5b7546af3",
  "a28d5e34007fd8fe6d7903044eb23a60fdad3c00",
  "9cabee3d27426676b852ce6b804cb2fdff7cd0b5",
  "ca5f6d76eab6e4f5aacee2864c79034d7111b986",
  "0ada6a2f6c8f891769366fc9aa6fd9f1facb36cf",
  "d3c8064d011ef853e3be506b95a045f41f78e72a",
  "70abf056bd92e8f77cfc551748fa54a4e3018d5f",
  "79f30c1cbc1b9c4949c8b85acc24a7578e01558b",
  "657f610728eef97d55e50212871f0993bb7cc700",
  "a0bc50b27e7ac39060ed015a55f2f4508c84f0c2",
  "65f927187bf96aac5d968fcc9351e5492b5f9356",
  "d2c9a93755aed20c4d8f55c1e92b812d2c7d49d2",
  "5fbd442a4647b079cda1229ecf6d8f44f361c8ca",
  "27843a60a1e044c1e6e3cf119fdf64eb2b3e0d94",
  "3deeb0a7426a028b435b4ddd8d35ac85cf567237",
  "328d0f546695c5fa02105deb055cf2801d9b8ba1",
  "abd2ed5ad35b3a0870aadae6dceacc988ba63895",
  "4e874bcfb6f5896fe9e5dab3b26f59b2e2a9c09b",
  "7d2c6781d873ed2ba7a87f46f735f5e15a41a6f1",
  "df0e187bb3895806261c87cf66e1772566ee8e58",
  "0903f4c5cb599a7928aef27e314e98291d1e3888",
  "61f396d2a4a13ab7a39ba791fac4b921b54a208e",
  "2159f75e5440c36431aedbc7dc29a65a327778b8",
  "2cf52c08cc76091d884e800c1c697b13f69907d4",
  "94146c9889748c7b727eb3ac8c20e53c52effd32",
  "6e12e4498ae69fb07ff2d8aab036a911229d6c62",
  "8b1c221a030cf720736d9fb7d0499dd7276fc1b3",
  "b11c5811ea074a30142d824b6e8cfd3df14b2688",
  "ffce8ce225c5d80098f50e877125b655aef6d101",
  "4b6126597881cb6ba1a45c1f6286769e7a094fb4",
  "0d2c377e936b68c70066613b10c0fdad537f90da",
  "6174515fa66c91bff1128913edd4e0f1de37cee0",
  "8eceffd5a41e678b6467c9bc80ce35d2e8543d98",
  "7d60db187337cbd881140d69d84c9246eda8382e",
  "246fba5c519576d93c5fac899c44b29b72f526ae",
  "a9c66694dd65b2fdf40cdf45a0c308cbd38004fc",
  "8074f0d02f748fc55448a4bf200f1dade8517059",
  "82cbc7b39d35af358f9e2513af13b2f77b647a00",
  "5ade564399e670bd1d429583059067f3a6ca2b7f",
  "9b42e4264020f3c3dcaaed806578ccd469205060",
  "d06fe6d3e5f7facb5855eca839422fe3824a5d6e",
  "3c80e7d9964a1c83a6dd5dc64e105e0e634bd58a",
  "03f6d90b7dbd0497dc3a52d1c27e23bb8c75897f",
  "4fd8cc4e27a3489b57ed986efe3d327d3de40d92",
  "31fe789b43277e35ab410f2afcfb574280af2dd8",
  "fa526a1611ccc66dec815cb963174118074b736e",
  "dcb5bf35b6d521195e613c42483f520139e2331d",
  "b76a0f1819c4be0a1482567ca9b9fbed3eda444c",
  "c3ab2ac23512d9bf62b02775e22cf80df814eb1b",
  "4f4422eb61c45edb4d76f10cd871c9f983f2ebaa",
  "e9355152c117c9e1fad8be86e9abea961cef4a36",
  "b485becd88db1ab3d556d405204451ba00adaa7d",
  "308a95d1d3f7bb556f48da4f4344566e59f6f1cb",
  "b7580969dd56151f608931f126f793bbf45d8fa0",
  "8ac800124d5b16afd57b5cf7245edfcd5885ea3b",
  "f730cb929a1c45032387c345e10d2427bea55a5e",
  "290abc4a2244bf0ecf5aa1ccee8ac8f60f8bce48",
  "463e7e879b7bdc6a97ec02a2a603aa1a46a04c80",
  "5cd1d0ccf98f2bd5a4bfaa770d55f16498af0bcc",
  "e2950a9f07b44e724df2129360cc140293c08308",
  "ab0041a7d3f7b2c290b5b834d46bdc7b7eb85815",
  "a6f6a0543ae470c6b056021cb2ac153368eafeec",
  "8506f3c10f75044946f1a23a7caf578253649471",
  "d5bd2050b74b309d54819ca17add173c6fca1e16",
  "8459b8870fcefff59f172d716b7bfe9fcc28d408",
  "89b144ecfdd5ea352083bf624d3cf842ec06a5e3",
  "be89f4e983dfb04e2b58a12eb9ed18149e108b07",
  "d5c4e380b1ec2f7f2068cfba9a90cb3ae7816110",
  "5853a3f24990deecced49d6bc15990102ec0c33a",
  "cb579ee537b34926d47129a0b54c0e6d00ef3004",
  "a0620472cde03e77caece701ab7bc5928a5d367c",
  "d5e1f52cbe4a11a3730b98f52109b57602a9c4a1",
  "aebd99f07218109162a905d0e0c91e58bedc83c5",
  "d8e611bafd70a549f035cf61ab0d6ed9e7f25c4e",
  "0f33a5033b750e6c4dca47e38ba020e912df143e",
  "8722453383f781d5427a4ee211020e49bf34a2b9",
  "bd175729d4177259c71cf13fd4ecfb5d01542706",
  "6330fd8ae91df4a5d7fbc2390c182ec6676dc5a6",
  "3b3e137b1bec6f20c9a8b2ad4f5784661fb0fa79",
  "3de95e18f18a54e2269bbf8f1a38ea70762c73fa",
  "c697b620c7c4015e32dd7bdd7d0430b33404e107",
  "d0a0e45b950e3871d8783b973409042b4ab382d4",
  "9d0149b0962d44bfc08a9f64d5afceb6281d7fb5",
  "24c130eb6cc0d8f663a8f6d16ffc61f935a2e02e",
  "1ac73bff74924ad9b74236c4962be27174ae87d0",
  "3b96d8565569421f43684b2c4eaa0639cbb5e011",
  "b56c55b9a70c8e2f07979b862374aed0e92a6dda",
  "b543e2e592200beb38235f6e48f8abe1d87ad872",
  "6b9895c31bcdb2d9c929b9da7e389ed91de672a0",
]
`;
